lnrpc/routerrpc/subscribe_events.go

```diff
@@ -0,0 +1,214 @@
+// +build routerrpc
+
+package routerrpc
+
+import (
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/htlcswitch"
+	"github.com/lightningnetwork/lnd/invoices"
+)
+
+// rpcHtlcEvent returns a rpc htlc event from a htlcswitch event.
+func rpcHtlcEvent(htlcEvent htlcswitch.HtlcEvent) (*HtlcEvent, error) {
+	key := htlcEvent.Key()
+
+	rpcEvent := &HtlcEvent{
+		IncomingChannelId: key.IncomingCircuit.ChanID.ToUint64(),
+		OutgoingChannelId: key.OutgoingCircuit.ChanID.ToUint64(),
+		IncomingHtlcId:    key.IncomingCircuit.HtlcID,
+		OutgoingHtlcId:    key.OutgoingCircuit.HtlcID,
+		Timestamp:         uint64(htlcEvent.Timestamp().Unix()),
+	}
+
+	switch e := htlcEvent.(type) {
```

Comment:
If you have this switch here anyway, do you need the `HtlcEvent` interface?

Reply:
Just a lot more lines of duplicate code to have to extract the htlc key and timestamp for each event type when it can be done all in one.

I think the interface is also generally nice to have as a prescription for what a htlc event should have; eg if something like a htlc interceptor/forwarding predicate goes in, we may want to add another event, which would need to fit this interface.

Reply:
I don't think it is more duplicated code, the extraction just happens in the rpc server instead of in htlcnotifier.

My general thought is that if we need to flatten events just for the rpc (at the moment), it should happen as close to the rpc code as possible.

I quickly tried it out: https://github.com/lightningnetwork/lnd/commit/f7ff5eb0f238a764b5efcb37eb191e642c953209

It is not a blocker if you like the interface more.

Reply:
Fair enough, actually less code relative to adding the interface. Fewer changes always better :) 

Cut the interface and flattened in rpc.